src/lib.rs

```diff
@@ -150,13 +212,68 @@ impl<'object, Endian> DebugInfo<'object, Endian>
         })
     }
 
-    pub fn locate(&self, addr: u64) -> Option<(String, u64)> {
+    pub fn locate(&self, addr: u64) -> Option<(String, u64, Option<Cow<str>>)> {
+        let mut func: Option<(&FunctionUnit, &gimli::Range, u64)> = None;
+        let mut loc = None;
         let mut current: Option<gimli::LineNumberRow> = None;
-        for unit in &self.units {
+        'outer: for unit in &self.units {
             if !unit.contains_address(addr) {
                 continue;
             }
 
+            if let Target::Function(ref u) = *unit {
+                let (range, dist) = u.ranges
+                    .iter()
+                    .filter(|range| addr >= range.begin && addr < range.end)
+                    .map(|range| (range, addr - range.begin))
+                    .min_by_key(|&(_, dist)| dist)
+                    .unwrap();
+
+                if let Some((prev, prange, pdist)) = func.take() {
+                    // are we a better match?
+                    func = if dist == pdist {
+                        // we're equally good -- are we tighter?
+                        if range.end < prange.end {
+                            // yes!
+                            Some((u, range, dist))
+                        } else if range.end == prange.end {
+                            println!("found two equivalent ranges for {:x}: {} and {}",
+                                     addr,
+                                     prev.name.to_string_lossy(),
+                                     u.name.to_string_lossy());
```

Comment:
Does this ever happen? Probably should use one of the logging crates for messages like this in the library code, instead of printing to stdout.